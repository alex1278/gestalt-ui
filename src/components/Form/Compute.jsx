import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Field } from 'react-final-form';
import { Row, Col } from 'react-flexybox';
import { TextField } from 'components/ReduxFormFields';
import { Slider, FontIcon } from 'react-md';
import { composeValidators, fixInputNumber, fixInputDecimal, min, max, mod, required } from 'util/forms';

class ComputeForm extends Component {
  static propTypes = {
    formValues: PropTypes.object.isRequired,
    form: PropTypes.object.isRequired,
    cpuName: PropTypes.string,
    memName: PropTypes.string,
    cpuMin: PropTypes.number,
    cpuMax: PropTypes.number,
    cpuStep: PropTypes.number,
    cpuStepSlider: PropTypes.number,
    memMin: PropTypes.number,
    memMax: PropTypes.number,
    memStep: PropTypes.number,
    memStepSlider: PropTypes.number,
    cpuDiscrete: PropTypes.bool,
    memDiscrete: PropTypes.bool,
  };

  static defaultProps = {
    cpuName: 'properties.cpus',
    memName: 'properties.memory',
    cpuMin: 0.1,
    cpuMax: 4.0,
    cpuStep: 0.1,
    cpuStepSlider: 0.1,
    memMin: 128,
    memMax: 2048,
    memStep: 1,
    memStepSlider: 32,
    cpuDiscrete: true,
    memDiscrete: true,
  };

  handleFocus = (event) => {
    const { target } = event;

    setTimeout(() => target.select());
  }

  onChangeCPU = (value) => {
    const { form, cpuName, cpuMin } = this.props;
    const formattedValue = value || value >= 0
      ? fixInputDecimal(value)
      : cpuMin;

    if (typeof formattedValue === 'number') {
      form.change(cpuName, formattedValue);
    }
  }

  onChangeMemory = (value) => {
    const { form, memName, memMin } = this.props;
    const formattedValue = value || value >= 0
      ? fixInputNumber(value)
      : memMin;

    if (typeof formattedValue === 'number') {
      form.change(memName, formattedValue);
    }
  }

  render() {
    const {
      formValues,
      cpuName,
      memName,
      cpuMin,
      cpuMax,
      cpuStep,
      cpuStepSlider,
      memMin,
      memMax,
      memStep,
      memStepSlider,
      cpuDiscrete,
      memDiscrete,
    } = this.props;

    return (
      <Row gutter={10}>
        <Col flex={10} xs={8}>
          <Slider
            id={`${cpuName}--cpu`}
            label="CPU"
            min={cpuMin}
            max={cpuMax}
            step={cpuStepSlider}
            leftIcon={<FontIcon>memory</FontIcon>}
            value={formValues.properties.cpus}
            onChange={this.onChangeCPU}
            discrete={cpuDiscrete}
            valuePrecision={1}
          />
        </Col>
        <Col flex={2} xs={4} alignSelf="center">
          <Field
            component={TextField}
            name={cpuName}
            step={cpuStep}
            placeholder="CPU"
            type="number"
            required
            onChange={this.onChangeCPU}
            onFocus={this.handleFocus}
            parse={fixInputDecimal}
            format={fixInputDecimal}
            validate={composeValidators(
              min(cpuMin),
              max(cpuMax),
              required('cpu is required'))
            }
          />
        </Col>
        <Col flex={10} xs={8}>
          <Slider
            id={`${memName}--memory`}
            label="Memory (MB)"
            min={memMin}
            max={memMax}
            step={memStepSlider}
            leftIcon={<FontIcon>sd_card</FontIcon>}
            value={formValues.properties.memory}
            onChange={this.onChangeMemory}
            discrete={memDiscrete}
            discreteTicks={memMin}
            tickWidth={6}
          />
        </Col>
        <Col flex={2} xs={4} alignSelf="center">
          <Field
            component={TextField}
            name={memName}
            step={memStep}
            placeholder="Memory"
            type="number"
            required
            onChange={this.onChangeMemory}
            onFocus={this.handleFocus}
            parse={fixInputNumber}
            format={fixInputNumber}
            validate={composeValidators(
              min(memMin),
              max(memMax),
              mod(memStep),
              required('memory is required'))
            }
          />
        </Col>
      </Row>
    );
  }
}

export default ComputeForm;
