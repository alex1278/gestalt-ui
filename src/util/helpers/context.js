/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
