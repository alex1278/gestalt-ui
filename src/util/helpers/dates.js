import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
