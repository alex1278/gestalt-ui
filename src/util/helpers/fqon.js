// sad hack for parsing url

export default function fqonParent(organization) {
  if (organization.org) {
    const urlParser = document.createElement('a');
    urlParser.href = organization.org.href;
    return urlParser.pathname.replace('/', '').split('/')[0];
  }

  return organization;
}
