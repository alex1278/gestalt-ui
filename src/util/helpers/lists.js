export function insertItem(array, payload) {
  return [...array, payload];
}

export function removeItem(array, payload) {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a === payload), 1);

  return newArray;
}

export function removeItemById(array, id, keyField = 'id') {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === id), 1);

  return newArray;
}

export function updateItem(array, payload) {
  return array.map((item) => {
    if (item.id === payload.id) {
      return {
        ...payload,
      };
    }

    // Leave every other item unchanged
    return item;
  });
}
