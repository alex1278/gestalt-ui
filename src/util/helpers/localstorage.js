/**
 * getItem
 * @param {*} key
 */
export const getItem = (key) => {
  try {
    return localStorage.getItem(key);
  } catch (err) {
    return undefined;
  }
};

/**
 * saveItem
 * @param {*} key
 * @param {*} value
 */
export const saveItem = (key, value) => {
  try {
    localStorage.setItem(key, value);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};

/**
 * removeItem
 * @param {*} key
 * @param {*} value
 */
export const removeItem = (key) => {
  try {
    localStorage.removeItem(key);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};
