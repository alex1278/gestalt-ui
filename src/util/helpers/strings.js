import { startCase, toLower } from 'lodash';

/**
 * getLastFromSplit
 * Parses and returns the last item in a structure like GrandPareng::Parent::Child
 * @param {String} string
 */
export function getLastFromSplit(string, delimiter = '::') {
  if (string) {
    const split = string.split(delimiter);

    return split[split.length - 1];
  }
  return string;
}

/**
 * truncate
 * @param {String} string
 * @param {Number} at
 * @param {String} trailing
 */
export function truncate(string, at, trailing = '...') {
  if (string) {
    if (string.length < at) {
      return string;
    }

    return `${string.substring(0, at).trim()}${trailing}`;
  }

  return string;
}

/**
 * toTitleCase
 * @param {String} string
 */
export function toTitleCase(string) {
  return startCase(toLower(string));
}


// TODO: Will be rmeoved when we have security config
export const checkIfPassword = field =>
  field && field.length && (field.toUpperCase().includes('PASSWORD') || field.toUpperCase().includes('SECRET') || field.toUpperCase().includes('KEY'));

export default {
  getLastFromSplit,
  truncate,
  toTitleCase,
  checkIfPassword
};

export const removeHostFromURL = (url) => {
  if (url) {
    return url.replace(/^.*\/\/[^/]+:?[0-9]?\//i, '');
  }

  return url;
};

export const isBase64 = (str) => {
  try {
    return btoa(atob(str)) === str;
  } catch (err) {
    return false;
  }
};
