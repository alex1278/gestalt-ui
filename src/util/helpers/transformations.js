import map from 'lodash/map';
import compact from 'lodash/compact';
import get from 'lodash/get';

/**
 * arrayToMap
 * @param {*} array
 * @param {*} keyName
 * @param {*} valueName
 */
export function arrayToMap(array = [], keyName = 'name', valueName = 'value') {
  if (!Array.isArray(array)) {
    return array;
  }

  return Object.assign({}, ...array.map(v => ({ [v[keyName]]: v[valueName] || '' })));
}

/**
 * mapTo2DArray
 * @param {*} object
 * @param {*} keyName
 * @param {*} valueName
 * @param {*} mergeSet
 */
export function mapTo2DArray(object = {}, keyName = 'name', valueName = 'value', mergeSet = {}) {
  if (typeof object !== 'object') {
    return object;
  }

  return map(object, (value, key) => (Object.assign({ [keyName]: key, [valueName]: value || '' }, mergeSet)));
}

/**
 * stringDemiltedToArray
 * convert comma delimited string to an array and remove blank entries
 * @param {*} string
 */
export function stringDemiltedToArray(string) {
  return compact(string.split(','));
}

/**
 * convertFromMaps
 * used to convert env vars from a map to an array and reconcile inherited vs own
 * @param {*} own
 * @param {*} inherited
 */
export function convertFromMaps(own = {}, inherited = {}, keyName = 'name', valueName = 'value') {
  const ownVars = mapTo2DArray(own, keyName, valueName, { inherited: false });

  // base case just return the mapped own vars
  if (!Object.keys(inherited)) {
    return ownVars;
  }

  const inheritedVars = mapTo2DArray(inherited, keyName, valueName, { inherited: true, overridden: false });

  const reconcileInheritedVars = inheritedVars.map((i) => {
    // find the dupe value in ownVars
    const index = ownVars.findIndex(o => o[keyName] === i[keyName]);

    // if the value is differnt mark the matching var as overriden
    if (index > -1 && ownVars[index][valueName] !== i[valueName]) {
      return { ...i, [valueName]: ownVars[index][valueName], inherited: true, overridden: true };
    }

    // default case the own env var
    return i;
  });

  const results = [...reconcileInheritedVars];

  // "merge" the ownVars back in
  ownVars.forEach((v) => {
    const index = inheritedVars.findIndex(o => o[keyName] === v[keyName]);

    if (index === -1) {
      results.push(v);
    }
  });

  return results;
}

/**
 * Builds an object from a delimited string
 * @param {String} string
 * @param {String} delimiter
 */
export function nestedObjectFromString(string, value = null, delimeter = '.') {
  const [last, ...paths] = string.split(delimeter).reverse();
  return paths.reduce((acc, el) => ({ [el]: acc }), { [last]: value });
}


export const getRoutePattern = (object, urlPattern) => {
  const keys = [];

  urlPattern.split('/').forEach((item) => {
    const search = item.startsWith(':');

    if (search) {
      keys.push(get(object, item.slice(1)));
    } else {
      keys.push(item);
    }
  });

  return keys.join('/');
};
