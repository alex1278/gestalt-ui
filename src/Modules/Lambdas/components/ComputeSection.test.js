import React from 'react';
import { Form } from 'react-final-form';
import Compute from './ComputeSection';
import { renderWithTheme } from '../../../../test/helpers';

describe('Lambdas::Components::ComputeSection', () => {
  it('mounts with basic props', () => {
    const formValuesMock = {
      properties: {
        cpus: 1,
        memory: 256,
      }
    };

    const { container } = renderWithTheme(
      <Form
        initialValues={formValuesMock}
        render={() => <Compute form={{ change: jest.fn() }} />}
        onSubmit={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
