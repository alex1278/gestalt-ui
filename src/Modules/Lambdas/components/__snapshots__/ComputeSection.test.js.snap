// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambdas::Components::ComputeSection mounts with basic props 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

.c1 {
  box-sizing: border-box;
  margin: 5px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  --width: calc(83.33333333333334% - 10px);
  -webkit-flex: 0 1 var(--width);
  -ms-flex: 0 1 var(--width);
  flex: 0 1 var(--width);
  min-width: var(--width);
}

.c3 {
  box-sizing: border-box;
  margin: 5px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  --width: calc(16.666666666666664% - 10px);
  -webkit-flex: 0 1 var(--width);
  -ms-flex: 0 1 var(--width);
  flex: 0 1 var(--width);
  min-width: var(--width);
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 .md-slider-thumb {
  z-index: 2;
}

.c2 .md-slider-discrete-tick {
  z-index: 1;
}

@media (min-width:0) and (max-width:599px) {
  .c1 {
    --width: calc(66.66666666666666% - 10px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

@media (min-width:0) and (max-width:599px) {
  .c3 {
    --width: calc(33.33333333333333% - 10px);
    -webkit-flex: 0 1 var(--width);
    -ms-flex: 0 1 var(--width);
    flex: 0 1 var(--width);
    min-width: var(--width);
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="md-slider-container c2 md-pointer--hover"
    >
      <label
        class="md-slider-label md-text"
        for="properties.cpus--cpu"
      >
        CPU
      </label>
      <input
        class="md-slider-input"
        id="properties.cpus--cpu"
        max="4"
        min="0.1"
        readonly=""
        type="range"
        value="1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-1 MuiSvgIcon-colorAction-4 MuiSvgIcon-fontSizeSmall-8"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15 9H9v6h6V9zm-2 4h-2v-2h2v2zm8-2V9h-2V7c0-1.1-.9-2-2-2h-2V3h-2v2h-2V3H9v2H7c-1.1 0-2 .9-2 2v2H3v2h2v2H3v2h2v2c0 1.1.9 2 2 2h2v2h2v-2h2v2h2v-2h2c1.1 0 2-.9 2-2v-2h2v-2h-2v-2h2zm-4 6H7V7h10v10z"
        />
      </svg>
      <div
        class="md-slider-track md-slider-track--ind-left"
      >
        <hr
          class="md-slider-track-fill"
          style="width: 23.076923076923077%;"
        />
        <div
          aria-pressed="false"
          class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-slider-thumb md-slider-thumb--on md-slider-thumb--discrete"
          role="button"
          tabindex="0"
        />
        <span
          class="md-slider-thumb md-slider-thumb--mask"
        />
      </div>
    </div>
  </div>
  <div
    class="c3"
  >
    <div
      class="MuiFormControl-root-10 MuiFormControl-marginDense-12 MuiFormControl-fullWidth-13"
      step="0.1"
    >
      <label
        class="MuiFormLabel-root-25 MuiFormLabel-filled-29 MuiFormLabel-required-30 MuiInputLabel-required-18 MuiInputLabel-root-14 MuiInputLabel-formControl-19 MuiInputLabel-animated-22 MuiInputLabel-shrink-21 MuiInputLabel-marginDense-20 MuiInputLabel-outlined-24"
        data-shrink="true"
      >
        CPU
        <span
          class="MuiFormLabel-asterisk-31"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root-45 MuiOutlinedInput-root-32 MuiInputBase-fullWidth-54 MuiInputBase-formControl-46 MuiInputBase-marginDense-52"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-62 MuiOutlinedInput-notchedOutline-39"
          style="padding-left: 8px;"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-63"
            style="width: 0px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
        <input
          aria-invalid="false"
          class="MuiInputBase-input-55 MuiOutlinedInput-input-40 MuiInputBase-inputType-58 MuiInputBase-inputMarginDense-56 MuiOutlinedInput-inputMarginDense-41"
          name="properties.cpus"
          required=""
          type="number"
          value="1"
        />
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="md-slider-container c2 md-pointer--hover"
    >
      <label
        class="md-slider-label md-text"
        for="properties.memory--memory"
      >
        Memory (MB)
      </label>
      <input
        class="md-slider-input"
        id="properties.memory--memory"
        max="2048"
        min="128"
        readonly=""
        type="range"
        value="256"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-1 MuiSvgIcon-colorAction-4 MuiSvgIcon-fontSizeSmall-8"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M18 2h-8L4.02 8 4 20c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-6 6h-2V4h2v4zm3 0h-2V4h2v4zm3 0h-2V4h2v4z"
        />
      </svg>
      <div
        class="md-slider-track md-slider-track--ind-left"
      >
        <span
          class="md-slider-discrete-tick"
          style="left: 0%; width: 3px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="width: 6px;"
        />
        <span
          class="md-slider-discrete-tick"
          style="left: 100%; width: 3px;"
        />
        <hr
          class="md-slider-track-fill"
          style="width: 6.666666666666667%;"
        />
        <div
          aria-pressed="false"
          class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-slider-thumb md-slider-thumb--on md-slider-thumb--discrete"
          role="button"
          tabindex="0"
        />
        <span
          class="md-slider-thumb md-slider-thumb--mask"
        />
      </div>
    </div>
  </div>
  <div
    class="c3"
  >
    <div
      class="MuiFormControl-root-10 MuiFormControl-marginDense-12 MuiFormControl-fullWidth-13"
      step="1"
    >
      <label
        class="MuiFormLabel-root-25 MuiFormLabel-filled-29 MuiFormLabel-required-30 MuiInputLabel-required-18 MuiInputLabel-root-14 MuiInputLabel-formControl-19 MuiInputLabel-animated-22 MuiInputLabel-shrink-21 MuiInputLabel-marginDense-20 MuiInputLabel-outlined-24"
        data-shrink="true"
      >
        Memory
        <span
          class="MuiFormLabel-asterisk-31"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root-45 MuiOutlinedInput-root-32 MuiInputBase-fullWidth-54 MuiInputBase-formControl-46 MuiInputBase-marginDense-52"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-62 MuiOutlinedInput-notchedOutline-39"
          style="padding-left: 8px;"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-63"
            style="width: 0px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
        <input
          aria-invalid="false"
          class="MuiInputBase-input-55 MuiOutlinedInput-input-40 MuiInputBase-inputType-58 MuiInputBase-inputMarginDense-56 MuiOutlinedInput-inputMarginDense-41"
          name="properties.memory"
          required=""
          type="number"
          value="256"
        />
      </div>
    </div>
  </div>
</div>
`;
